### PR TITLE
mgr/cephadm: Add CephadmDaemonSpec class 

### PR DESCRIPTION
**Pull request by sebastian-philipp**
This is fallout from #35456 while trying to figure out how to match exceptions with the current subject.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
todo:

```
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 107, in do_work
    return f(self, *arg)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1369, in _daemon_actions
    return self._daemon_action(daemon_type, daemon_id, host, action)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1380, in _daemon_action
    return self._create_daemon(daemon_spec)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1668, in _create_daemon
    self.cache.update_daemon_config_deps(daemon_spec.host, daemon_spec.name(), deps, start_time)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/inventory.py", line 242, in update_daemon_config_deps
    'last_config': stamp,
KeyError: 'myrgw.foobar'
ERROR    orchestrator._interface:_interface.py:321 _Promise failed
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/orchestrator/_interface.py", line 274, in _finalize
    next_result = self._on_complete(self._value)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 132, in <lambda>
    return CephadmCompletion(on_complete=lambda _: f(*args, **kwargs))
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1413, in daemon_action
    return self._daemon_actions(args)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 114, in forall_hosts_wrapper
    return CephadmOrchestrator.instance._worker_pool.map(do_work, vals)
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 266, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 107, in do_work
    return f(self, *arg)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1369, in _daemon_actions
    return self._daemon_action(daemon_type, daemon_id, host, action)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1380, in _daemon_action
    return self._create_daemon(daemon_spec)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/module.py", line 1668, in _create_daemon
    self.cache.update_daemon_config_deps(daemon_spec.host, daemon_spec.name(), deps, start_time)
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/cephadm/inventory.py", line 242, in update_daemon_config_deps
    'last_config': stamp,
KeyError: 'myrgw.foobar'
```

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-06-23_07:49:50-rados-wip-kefu-testing-2020-06-23-1007-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/

---

- http://pulpito.ceph.com/kchai-2020-06-23_07:49:50-rados-wip-kefu-testing-2020-06-23-1007-distro-basic-smithi/5171960, no clues, osd.2 failed to reply queries from osd.0 when peering. but there was no clues in its log.
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/5173648:  tracked by https://tracker.ceph.com/issues/45721
- http://pulpito.ceph.com/kchai-2020-06-23_15:15:59-rados-wip-kefu-testing-2020-06-23-1353-distro-basic-smithi/5173654/: no tests failed..

**Comment by tchaikov** [comment]
jenkins test docs

**Comment by sebastian-philipp** [comment]
jenkins retest this please

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 1]
```diff
@@ -14,6 +14,48 @@
 
 logger = logging.getLogger(__name__)
 
+ServiceSpecs = TypeVar('ServiceSpecs', bound=ServiceSpec)
+
+
+class CephadmDaemonSpec(Generic[ServiceSpecs]):
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:20]: I would like to have a generic  <ContainerSpec>, and then a <CephadmDaemonSpec> with  a <ContainerSpec> attribute.

Things to have in the generic "ContainerSpec": ( the kind of things we need to launch a container):
I think we should have all the things that can be used in a `podman run command`... but a good start will be:
image: The image to use for the container
config_file: location + content
mounts: storage in the host
ports: ports to open
name: container name
network: network mode for the container
command: command to execute

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 2]
```diff
@@ -14,6 +14,48 @@
...
+        # For run_cephadm. Would be great to have more expressive names.
+        self.extra_args: List[str] = extra_args or []
+        self.extra_config: Dict[str, Any] = extra_config or {}
+
+
+    def name(self) -> str:
+        return '%s.%s' % (self.daemon_type, self.daemon_id)
+
 
 class CephadmService(metaclass=ABCMeta):
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:60]: CephadmService class ... This abstraction seems not to be for a service, it seems to be a "factory" class for create daemons.... In my opinion the functionality included here should be moved to  a "CephadmDaemon" class

The CephadmService class should model a "Service"... and in a Service what we have is:
- one or more daemons providing the service
- health of the service
- tools to scale up/down the service
- tools to provide a virtual ip ( for load balancing and HA)

**Comment by jmolmo** [comment]
Still not the thing we need, but a giant step in the right direction: :-)

I would like something like this class set :
container: Abstraction of a container with all the  things we need to start a container in one host
daemon: A process running as a systemd daemon (uses container)
cephdaemon: A ceph process running in a ceph cluster (uses daemon)
cephService: a set of cephdaemons using "service specs"

And a set of "service specs" templates to define a Ceph Cluster

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-07-03_08:12:34-rados:cephadm-wip-swagner-testing-2020-07-02-1034-distro-basic-smithi/

* https://tracker.ceph.com/issues/46336

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/cephadm/services/cephadmservice.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:20]: hm. is this something for cephadm or for mgr/cephadm? right now, this code lives in the binary.

File: src/pybind/mgr/cephadm/services/cephadmservice.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:60]: hm that's not totally correct. Like e.g. `get_active_daemon` which returns the active daemon for a set of daemons belonging to a service. But sure, some methods work on daemons belonging to a service.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-07-15_10:39:40-rados:cephadm-wip-swagner-testing-2020-07-15-1019-distro-basic-smithi/

**Pull request merged**